src/Common/src/System/Net/SafeCloseSocket.Unix.cs

```diff
@@ -0,0 +1,213 @@
+
+            public SocketAsyncContext AsyncContext
+            {
+                get
+                {
+                    if (Volatile.Read(ref _asyncContext) == null)
+                    {
+                        Interlocked.CompareExchange(ref _asyncContext, new SocketAsyncContext((int)handle, SocketAsyncEngine.Instance), null);
+                    }
+                    return _asyncContext;
+                }
+            }
+
+            private unsafe SocketError InnerReleaseHandle()
+            {
+				int errorCode;
+
+                // If m_Blockable was set in BlockingRelease, it's safe to block here, which means
+                // we can honor the linger options set on the socket.  It also means closesocket() might return WSAEWOULDBLOCK, in which
+                // case we need to do some recovery.
+                if (_blockable)
+                {
+                    GlobalLog.Print("SafeCloseSocket::ReleaseHandle(handle:" + handle.ToString("x") + ") Following 'blockable' branch.");
+
+					errorCode = Interop.Sys.Close((int)handle);
+					if (errorCode == -1)
+					{
+						errorCode = (int)Interop.Sys.GetLastError();
+					}
+                    GlobalLog.Print("SafeCloseSocket::ReleaseHandle(handle:" + handle.ToString("x") + ") close()#1:" + errorCode.ToString());
+
+#if DEBUG
+                    _closeSocketHandle = handle;
+                    _closeSocketResult = SocketPal.GetSocketErrorForErrorCode((Interop.Error)errorCode);
+#endif
+
+                    // If it's not EWOULDBLOCK, there's no more recourse - we either succeeded or failed.
+					if (errorCode != (int)Interop.Error.EWOULDBLOCK)
+                    {
+                        if (errorCode == 0)
+                        {
+                            _asyncContext.Close();
+                        }
+						return SocketPal.GetSocketErrorForErrorCode((Interop.Error)errorCode);
+                    }
+
+                    // The socket must be non-blocking with a linger timeout set.
+                    // We have to set the socket to blocking.
+                    errorCode = Interop.Sys.Fcntl.SetIsNonBlocking((int)handle, 0);
+                    if (errorCode == 0)
+                    {
+                        // The socket successfully made blocking; retry the close().
+                        errorCode = Interop.Sys.Close((int)handle);
+
+                        GlobalLog.Print("SafeCloseSocket::ReleaseHandle(handle:" + handle.ToString("x") + ") close()#2:" + errorCode.ToString());
+#if DEBUG
+                        _closeSocketHandle = handle;
+                        _closeSocketResult = SocketPal.GetSocketErrorForErrorCode((Interop.Error)errorCode);
+#endif
+                        if (errorCode == 0)
+                        {
+                            _asyncContext.Close();
+                        }
+                        return SocketPal.GetSocketErrorForErrorCode((Interop.Error)errorCode);
```

Comment:
This whole if block looks identical to the earlier code... consider refactoring it out into a helper method.